### PR TITLE
added customizable pipeline tool

### PR DESCRIPTION
**Pull request by hanzhichao2000**
## Description

added a pipeline tool consisting multiple steps that makes tool/chain reuse easier.
with more generic tools reused by domain task tools, metadata-driven agent-tool chain loading will be easier.

##  Dependencies

None

## Maintainer

 - Agents / Tools / Toolkits: @vowelparrot

### CODE REVIEW

File: langchain/tools/pipeline/tool.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        out = input_next
+        out = _return_based_on_expression(self.output_expression, out, final=True)
+        return json.dumps(out)
+
+    async def _arun(
+        self,
+        tool_input: Union[str, Dict],
+        run_manager: Optional[AsyncCallbackManagerForToolRun] = None,
+    ) -> str:
+        return self._run(tool_input)
```
Comment (nfcampos) [inline langchain/tools/pipeline/tool.py:132]: this is not ideal

File: langchain/tools/pipeline/tool.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    within each step of the pipeline.
+    """
+
+    name = "pipeline_tool"
+    description = """
+    Can be used to construct a context task handler with existing tools and chains.
+    """
+
+    steps: List[PipelineStep]
+    output_expression: Optional[Union[str, List[str], Dict[str, str]]] = None
```
Comment (nfcampos) [inline langchain/tools/pipeline/tool.py:109]: Could you clarify what is the purpose of output_expression? And could it be implemented as just another step in the pipeline? If so I'd be inclined to remove it

File: langchain/tools/pipeline/tool.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            else:
+                kw[self.input_expression] = input_[self.input_expression]
+        if not self.input_expression:
+            args, kw = [input_], {}
+
+        out = self.func(*args, **kw)
+        return _return_based_on_expression(self.output_expression, out, extra=input_)
+
+
+class PipelineTool(BaseTool):
```
Comment (nfcampos) [inline langchain/tools/pipeline/tool.py:96]: Not sure if subclassing basetool is ideal here, I think we may want it to subclass StructuredTool instead? @hwchase17

File: langchain/tools/pipeline/tool.py (hunk 1 above)
Comment (hanzhichao2000) [reply langchain/tools/pipeline/tool.py]: @nfcampos thanks for your comment.  I have utilized a similar implementation, which can be found at https://github.com/hwchase17/langchain/blob/v0.0.221/langchain/tools/shell/tool.py#L75-L83

File: langchain/tools/pipeline/tool.py (hunk 3 above)
Comment (hanzhichao2000) [reply langchain/tools/pipeline/tool.py:97]: As a newcomer to Langchain contribution, I am not certainly sure about using the ``BaseTool`` class and its suitability for my needs. From my understanding, the ``StructuredTool`` class is specifically intended for transforming a generic function with multiple inputs into a tool, while the "PipelineTool" class is more geared towards constructing a predefined workflow. After reviewing the "StructuredTool" class, I have found limited reusable components that would be applicable to the ``PipelineTool``.

**Comment by hwchase17** [comment]
love this idea, few questions on implementation

biggest one: @nfcampos wdyt about changing this to just use sequential chain? that would let it log intermediate steps properly, and we can make it easy to go from arbitrary chain to tool (if we don't already).

**Comment by hanzhichao2000** [comment]
> love this idea, few questions on implementation
> 
> biggest one: @nfcampos wdyt about changing this to just use sequential chain? that would let it log intermediate steps properly, and we can make it easy to go from arbitrary chain to tool (if we don't already).

i got the pr messed and i will raise another pr the the discussion.

**Pull request closed without merging**